engine.py

```diff
@@ -209,11 +210,62 @@ def get_king_moves(self, r, c, moves):
                 destination = self.board[end_row][end_col]
                 if destination[1] == enemy_color or destination == "  ":
                     moves.append(Move((r, c), (end_row, end_col), self.board))
-        self.get_castle_moves(r, c, moves, enemy_color)
 
 
-    def get_castle_moves(self, r, c, moves, enemy_color):
+    def get_castle_moves(self, r, c, moves):
+        '''
+        Generate all valid castle moves for king and add to list of moves
+        '''
+        #condition no.1 for a king to castle
+        if self.square_under_attack(r, c, moves):
+            return # can't castle when in check
+
+        #condition no.2 for a king to castle: if squares on the king castling side or queen castling side are free
+        if (self.light_to_move and self.current_castling_right.lks) or (not self.light_to_move and self.current_castling_right.dks):
+            self.get_king_side_castle_moves(r, c, moves)
         
+        if (self.light_to_move and self.current_castling_right.lqs) or (not self.light_to_move and self.current_castling_right.dqs):
+            self.queen_side_castle_moves(r, c, moves)
+        
+    
+    def get_king_side_castle_moves(self, r, c, moves):
+        '''
+        This function checks whether the squares for a king side castle are empty and not under attack
+        '''
+        if self.board[r][c+1] == "  " and self.board[r][c+2] == "  ":
+            if not self.square_under_attack(r, c+1) and not self.square_under_attack(r, c+2):
+                moves.append(Move(r, c), (end_row, end_col), self.board, is_castle_move = True)
+
+
+    def queen_side_castle_moves(self, r, c, moves):
+        '''
+        This function checks whether the squares on the queen side castle are empty and not under attack
+        '''
+        if self.board[r][c-1] == "  " and self.board[r][c-2] == "  " and self.board[r][c-3] == "  ":
+            if not self.square_under_attack(r, c-1) and not self.square_under_attack(r, c-2):
+                moves.append(Move(r, c), (end_row, end_col), self.board, is_castle_move = True)
+
+    def king_in_check(self):
+
+        '''
+        determine if current player is in check
+        '''
+        if self.light_to_move:
+            return self.square_under_attack(self.light_king_location[0], self.light_king_location[1])
+        else:
+            return self.square_under_attack(self.dark_king_location[0], self.dark_king_location[1])
+
+    def square_under_attack(self, r, c, moves):
+        '''
+        determine if enemy can attack the square
+        '''
+        self.light_to_move = not self.light_to_move # switch turns
+        opp_moves = self.get_possible_moves()
+        self.light_to_move = not self.light_to_move # switch back turns
+        for move in opp_moves:
+            if move.end_row == r and move.end_col == c: # square is under attack
+                return True
+        return False
```

Comment:
You don't need to pass moves to this method, since you did not use it anywhere in the method